optuna/samplers/_search_space/group_decomposed.py

```diff
@@ -0,0 +1,103 @@
+import copy
+from typing import Dict
+from typing import List
+from typing import Optional
+
+from optuna.distributions import BaseDistribution
+from optuna.study import BaseStudy
+from optuna.trial import TrialState
+
+
+class _SearchSpaceGroup(object):
+    def __init__(self) -> None:
+        self._group: List[Dict[str, BaseDistribution]] = []
+
+    @property
+    def group(self) -> List[Dict[str, BaseDistribution]]:
+        return self._group
+
+    def add_distributions(self, distributions: Dict[str, BaseDistribution]) -> None:
+        self._group = _add_distributions(self.group, distributions)
+
+
+def _add_distributions(
+    group: List[Dict[str, BaseDistribution]], distributions: Dict[str, BaseDistribution]
+) -> List[Dict[str, BaseDistribution]]:
+    if len(distributions) == 0:
+        return group
+
+    for search_space in group:
+        _keys = set(search_space.keys())
+        _dist_keys = set(distributions.keys())
+
+        if _keys.isdisjoint(_dist_keys):
+            continue
+
+        if _keys < _dist_keys:
+            return _add_distributions(
+                group, {name: distributions[name] for name in _dist_keys - _keys}
+            )
+
+        if _keys > _dist_keys:
+            group.append(distributions)
+            group.append({name: search_space[name] for name in _keys - _dist_keys})
+            group.remove(search_space)
+            return group
+
+        _intersection = _keys & _dist_keys
+        group.append({name: search_space[name] for name in _intersection})
+        if len(_keys - _intersection) > 0:
+            group.append({name: search_space[name] for name in _keys - _intersection})
+        group.remove(search_space)
+        return _add_distributions(
+            group, {name: distributions[name] for name in _dist_keys - _intersection}
+        )
+
+    group.append(distributions)
+
+    return group
+
+
+class _GroupDecomposedSearchSpace(object):
```

Comment:
Mostly curious but is an instance of this class intended to be used across trials (like the `IntersectionSearchSpace`)? I don't think that the cursor is needed in that case which would simplify the logic.

Reply:
Thanks for the insightful comment. The `_GroupDecomposedSearchSpace` class is instantiated in every call of `infer_relative_search_space`, so we can remove the `cursor`. I'll simplify the logic.

Reply:
I made TPESampler own the `_GroupDecomposedSearchSpace` so that it doesn't have to check every trial in each `infer_relatice_search_space` call. To achieve this, I decided to use the cursor again.